pootle/apps/pootle_language/views.py

```diff
@@ -41,11 +42,25 @@ def url_kwargs(self):
         return {"language_code": self.object.code}
 
     def get_object(self):
-        return get_object_or_404(
-            Language, code__iexact=self.kwargs["language_code"])
+        lang = Language.get_canonical(self.kwargs["language_code"])
+        if lang is None:
+            raise Http404
+        return lang
+
+    def get(self, *args, **kwargs):
+        self.object = self.get_object()
+        if self.object.code != kwargs["language_code"]:
+            return redirect(
+                self.url_pattern_name,
+                self.object.code,
+                permanent=True)
+        response = super(LanguageMixin, self).get(*args, **kwargs)
+        response.set_cookie('pootle-language', self.object.code)
```

Comment:
Good catch; as you say it should only happen in `LanguageBrowseView` to preserve previous behavior. If there's a will to change it that's a separate matter.